shared-bindings/ota/__init__.c

```diff
@@ -0,0 +1,127 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "shared-bindings/ota/__init__.h"
+
+//| """OTA Module
+//|
+//| The `ota` module implements over-the-air update.
+//|
+//| There are two identical ota partitions ota_0/1, these
+//| contain different firmware versions.
+//| Having two partitions enables rollback functionality.
+//|
+//| The two partitions are defined as boot partition and
+//| next-update partition. Calling `ota.flash()` writes the
+//| next-update partition.
+//|
+//| After the next-update partition is written a validation
+//| check is performed and on a successful validation this
+//| partition is set as the boot partition. On next reset,
+//| firmware will be loaded from this partition.
+//|
+//| Here is the sequence of commands to follow:
+//|
+//| .. code-block:: python
+//|
+//|     import ota
+//|
+//|     ota.flash(buffer, offset)
+//|     ota.finish()
+//| """
+//| ...
+//|
+
+//| def switch() -> None:
+//|     """Switches the boot partition. On next reset,
+//|        firmware will be loaded from the partition
+//|        just switched over to."""
+//|     ...
+//|
+STATIC mp_obj_t ota_switch(void) {
+    common_hal_ota_switch();
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_0(ota_switch_obj, ota_switch);
+
+//| def finish() -> None:
+//|     """Validates flashed firmware, sets next boot partition.
+//|         **Must be called** after the firmware has been
+//|         completely written into the flash using `ota.flash()`.
+//|
+//|         This is to be called only once per update."""
+//|     ...
+//|
+STATIC mp_obj_t ota_finish(void) {
+    common_hal_ota_finish();
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_0(ota_finish_obj, ota_finish);
+
+//| def flash(*buffer: ReadableBuffer, offset: int=0) -> None:
+//|     """Writes one of two OTA partitions at the given offset.
+//|
+//|     The default offset is 0. It is necessary to provide the
+//|     offset when writing in discontinous chunks.
+//|
+//|     This can be called multiple times when flashing the firmware
+//|     in small chunks."""
```

Comment:
I don't think auto-increment is a good idea. If it's chunks, then it's easy enough to track the offset yourself. It's idempotent then too.

Reply:
auto-increment is for continuous chunks... the offset property is only required when flashing in discontinuous chunks.

Reply:
Right, and I don't think it should auto-increment. Doing so creates some hidden state. When flashing chunks, always require an offset.

Reply:
I made offset a required parameter.